docs: refactor routing doc (#35566)
angular/angular: aio/content/guide/router.md

Notice that this redirect precedes the wildcard route.
Here, `path: ''` means to use the initial relative URL (`''`).

A redirect route requires a `pathMatch` property to tell the router how to match a URL to the path of a route.
For more details on `pathMatch` see [Spotlight on `pathMatch`](guide/router#pathmatch).

{@a creating-child-routes}

## Creating child routes

To create child routes, first add `<router-outlet>` to the component template that you want to have child routes.
This means you're adding a second `<router-outlet>` to your app, because it is in addition to the `<router-outlet>` in `AppComponent`.

Generate additional child components so the router has a view to render.
In this example, there are two additional child components, `child-a`, and `child-b`.
Here, `FirstComponent` has its own `<nav>` and `<router-outlet>` (in addition to the one in `AppComponent`).

```html
<h2>First Component</h2>

<!-- Links to the two child routes -->
<nav>
  <ul>
    <li><a routerLink="child-a">Child A</a></li>
    <li><a routerLink="child-b">Child B</a></li>
  </ul>
</nav>

<!-- Add router-outlet to template -->
<router-outlet></router-outlet>

```

> Configure the child routes in the routing module by adding a `children` array to the `first-component` route as follows:

Reviewer: ```suggestion
A child route is like any other route, in that it needs both a `path` and a `component`. The one difference is that you place child routes in a `children` array within the parent route.
```

<code-example header="AppRoutingModule (excerpt)">
const routes: Routes = [
  { path: 'first-component',
    component: FirstComponent, // this is the component with the <router-outlet> in the template
    children: [
      {
        path: 'child-a', // child route path
        component: ChildAComponent // child route component that the router renders
      },
      {
        path: 'child-b',
        component: ChildBComponent // another child route component that the router renders
      }
    ] },
</code-example>

> Child routes are like any other route, with the exception that you put them in a `children` array of an already existing route.

Reviewer: If you go with my previous suggestion, you don't need lines 341 and 342.

Like other routes, you declare a `path` for the URL as well as a `component` for the router to render in the browser.

{@a using-relative-paths}

> ## Using relative paths

Reviewer: We don't talk about `relativeTo` here. Should we?

          this.name = params['name'];
          });
        }
      </code-example>

    Note: The preceding example uses a variable, `id`, and assigns it the value based on the `id` parameter.

{@a wildcard-route-how-to}

##  Setting up wildcard routes

A well-functioning application should gracefully handle when users attempt to navigate to a part of your application that does not exist.
To add this functionality to your application, you set up a wildcard route.
The Angular router selects this route any time the requested URL doesn't match any router paths.

To set up a wildcard route, add the following code to your `routes` definition.

<code-example header="AppRoutingModule (excerpt)">

{ path: '**', component: <component-name> }

</code-example>

The two asterisks, `**`, indicate to Angular that this `routes` definition is a wildcard route.
For the component property, you can define any component in your application.
Common choices include an application-specific `PageNotFoundComponent`, which you can define to [display a 404 page](guide/router#404-page-how-to) to your users; or a redirect to your application's main component.

> A wildcard route is the last route because it matches any URL.

Author: Clarified. The explanation you're referencing is still there in the tutorial part. I've added this here though:
> For more detail on why order matters for routes, see [Wildcard routes](guide/router#wildcard).

This section is a high level quick reference so doesn't include the in-depth explanations, but you're right about its importance here. Does the link help?

> A wildcard route is the last route because it matches any URL.

Author: Resolving as we talked about these offline.

  { path: 'second-component', component: SecondComponent },
  { path: '**', component: PageNotFoundComponent },  // Wildcard route for a 404 page
];

</code-example>

The last route with the `path` of `**` is a wildcard route.
The router selects this route if the requested URL doesn't match any of the paths earlier in the list and sends the user to the `PageNotFoundComponent`.

## Setting up redirects

To set up a redirect, configure your routes as for a 404, but make the component property different page.

<code-example header="AppRoutingModule (excerpt)">

const routes: Routes = [
  { path: 'first-component', component: FirstComponent },
  { path: 'second-component', component: SecondComponent },
  { path: '',   redirectTo: '/first-component', pathMatch: 'full' }, // redirect to `first-component`
  { path: '**', component: FirstComponent },
];

</code-example>

In this example, the third route is a redirect so that the router defaults to the `first-component` route.
Notice that this redirect precedes the wildcard route.

A redirect route requires a `pathMatch` property to tell the router how to match a URL to the path of a route.

> Here, `path: ''` means to use the initial relative URL (`''`).

Author: Clarified/reorganized this and linked to the more in-depth section on `pathMatch`. (We're still in the high-level reference part of the doc.)

> Here, `path: ''` means to use the initial relative URL (`''`).

Author: Resolving as we talked about these offline.

ngOnInit() {
  this.heroes$ = this.route.paramMap.pipe(
    switchMap(params => {
      // (+) before `params.get()` turns the string into a number
      this.selectedId = +params.get('id');
      return this.service.getHeroes();
    })
  );
}

</code-example>

Next, in the component that you want to navigate to, import the following members.

<code-example header="Component 2 (excerpt)">

import { Router, ActivatedRoute, ParamMap } from '@angular/router';
import { Observable } from 'rxjs';

</code-example>

Inject `ActivatedRoute` and `Router` in the constructor of the component class so they are available to this component:

<code-example header="Component 2 (excerpt)">

  item$: Observable<Item>;

  constructor(
    private route: ActivatedRoute,
    private router: Router  ) {}

  ngOnInit() {
    let id = this.route.snapshot.paramMap.get('id');
    this.hero$ = this.service.getHero(id);
  }

  gotoItems(item: Item) {
    let heroId = item ? hero.id : null;
    // Pass along the item id if available
    // so that the HeroList component can select that item.
    this.router.navigate(['/heroes', { id: itemId }]);
  }

</code-example>

## Lazy loading modules

> Too lazy load Angular modules, use `loadchildren` (instead of `component`) in your `AppRoutingModule` `routes` configuration as follows:

Author: Resolving as we talked about these offline.

For this example, accept the default of `CSS`.

### Adding components for routing

To use the Angular router, an app needs to have at least two components so that it can navigate from one to the other. To create a component using the CLI, enter the following at the command line where `first` is the name of your component:

<code-example language="none" class="code-shell">
  ng generate component first
</code-example>

Repeat this step for a second component but give it a different name.
Here, the new name is `second`.

<code-example language="none" class="code-shell">
  ng generate component second
</code-example>

The CLI automatically appends `Component`, so if you were to write `first-component`, your component would be `FirstComponentComponent`.

{@a basics-base-href}

<div class="alert is-helpful">

  #### `<base href>`

  This guide works with a CLI-generated Angular app.
  If you are working manually, make sure that you have `<base href="/">` in the `<head>` of your index.html file.
  The following example assumes that the `app` folder is the application root, and uses `"/"`.

  <code-example>

>     <!doctype html>

Reviewer 2: We should really try not to have hard-coded code snippets in the docs.
Is there any reason this can't be extracted from an example app?

Same for all the other hard-coded snippets in the guide.

Author: Per offline conversation, I've switch most of them to come from an example app. The lazy loading and accessing fragments ones will be fixed in a follow up PR.

First, import the following members in the component you want to navigate from.

<code-example header="Component import statements (excerpt)">

import { ActivatedRoute } from '@angular/router';
import { Observable } from 'rxjs';
import { switchMap } from 'rxjs/operators';

</code-example>

Next inject the activated route service:

<code-example header="Component (excerpt)">
constructor(private route: ActivatedRoute) {}
</code-example>

Configure the class so that you have an observable, `heroes$`, a `selectedId` to hold the `id` number of the hero, and the heroes in the `ngOnInit()`, add the following code to get the `id` of the selected hero.
This code snippet assumes that you have a heroes list, a hero service, a function to get your heroes, and the HTML to render your list and details, just as in the Tour of Heroes example.

<code-example header="Component 1 (excerpt)">
heroes$: Observable<Hero[]>;
selectedId: number;
heroes = HEROES;

ngOnInit() {
  this.heroes$ = this.route.paramMap.pipe(
    switchMap(params => {

>       // (+) before `params.get()` turns the string into a number
>       this.selectedId = +params.get('id');

Reviewer 3: ```suggestion
      this.selectedId = Number(params.get('id'));
```
Let's change this - the current unary operator is forbidden in the internal Google style guide:

> Code must not use unary plus (+) to coerce strings to numbers. Parsing numbers can fail, has surprising corner cases, and can be a code smell (parsing at the wrong layer). A unary plus is too easy to miss in code reviews given this.

  }
)

</code-example>

### Preloading component data

You can preload component data so that all elements and data on a page render at the same time when the user activates a route. To preload component data, you can use a `resolver`.

#### Resolvers

Create a resolver service. With the CLI, the command to generate a service is as follows:

<code-example language="none" class="code-shell">
  ng generate service <service-name>
</code-example>

In your service, import the following router members, implement `Resolve`, and inject the `Router` service:

<code-example header="Resolver service (excerpt)">
import {
  Router, Resolve,
  RouterStateSnapshot,
  ActivatedRouteSnapshot
}

...

export class CrisisDetailResolverService implements Resolve<> {
  constructor(private router: Router) {}
}
</code-example>

 Additionally, add the `resolve()` method to your service that gets the data for the page.
 The following is an example that you would fill in with the appropriate logic for your app.

<code-example header="Resolver service (excerpt)">

resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<> | Observable<never> {
  // your logic goes here
}

> </code-example>

Reviewer 3: I think this can really be one code example. Adding the `resolve` method is required as part of implementing the `Resolve` interface. 

Also, you have many unused/unnecessary imports.

i.e.

In your service, import the following router members, implement Resolve, and inject the Router service:
```
import { Resolve } from '@angular/router';

...

export class CrisisDetailResolverService implements Resolve<> {
  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<> {
    // your logic goes here
  }
}
```

Import this resolver into your module's routing module.

<code-example header="Feature module's routing module (excerpt)">

import { YourResolverService }    from './your-resolver.service';

</code-example>

Add a `resolve` object to the component's `route` configuration.

<code-example header="Feature module's routing module (excerpt)">
{
  path: '/your-path',
  component: YourComponent,
  resolve: {
    crisis: YourResolverService
  }
}
</code-example>

In the component, use an observable to get the data from the `ActivatedRoute`.

<code-example header="Component (excerpt)">
ngOnInit() {
  this.route.data

>     .subscribe(your-parameters) => {

Reviewer 3: ```suggestion
    .subscribe((your-parameters) => {
```

      // your data-specific code goes here
    });
}
</code-example>

For more information with a working example, see the [routing tutorial section on preloading](guide/router#preloading-background-loading-of-feature-areas).

## Preventing unauthorized access

Use route guards to prevent users from navigating to parts of an app without authorization.
The following route guards are available in Angular:

* [`CanActivate`](api/router/CanActivate)
* [`CanActivateChild`](api/router/CanActivateChild)
* [`CanDeactivate`](api/router/CanDeactivate)
* [`Resolve`](api/router/Resolve)
* [`CanLoad`](api/router/CanLoad)

To use route guards, consider using component-less routes as this facilitates guarding child routes.

Create a service for your guard:

<code-example language="none" class="code-shell">
  ng generate guard your-guard
</code-example>

In your guard class, implement the guard you want to use.
The following example uses `CanActivate` to guard the route.

<code-example header="Component (excerpt)">
export class YourGuard implements CanActivate {
  canActivate(
    next: ActivatedRouteSnapshot,

>     state: RouterStateSnapshot): boolean {

Reviewer 3: ```suggestion
    state: RouterStateSnapshot): boolean|UrlTree {
```

* The `CanActivateChild` guard (checking child route access).
* The `CanDeactivate` guard (ask permission to discard unsaved changes).
* The `Resolve` guard (pre-fetching route data).

> * Lazy loading NgModules.

Author: Updated to use code font but have to make it singular to avoid backtick-s per style guide.

</code-tabs>

### Register `Router` and `Routes`

> In order to use the Router, you must first register the `RouterModule` from the `@angular/router` package.

Reviewer 3: ```suggestion
In order to use the `Router`, you must first register the `RouterModule` from the `@angular/router` package.
```

<code-example path="router/src/app/auth/auth.service.ts" header="src/app/auth/auth.service.ts (excerpt)"></code-example>

Although it doesn't actually log in, it has an `isLoggedIn` flag to tell you whether the user is authenticated.
Its `login()` method simulates an API call to an external service by returning an observable that resolves successfully after a short pause.
The `redirectUrl` property stores the URL that the user wanted to access so you can navigate to it after authentication.

<div class="alert is-helpful">

To keep things simple, this example redirects unauthenticated users to `/admin`.

</div>

> <!--  KW - what does it refer to? -->

Reviewer 3: I'm guessing 'it' is the `AuthService`?

Also reminder here to use

`return this.router.createUrlTree(['/login']);`
instead of
```
this.router.navigate();
return false;
```

Revise the `AuthGuard` to call it.

<code-example path="router/src/app/auth/auth.guard.2.ts" header="src/app/auth/auth.guard.ts (v2)"></code-example>

Notice that you *inject* the `AuthService` and the `Router` in the constructor.
You haven't provided the `AuthService` yet but it's good to know that you can inject helpful services into routing guards.

This guard returns a synchronous boolean result.
If the user is logged in, it returns true and the navigation continues.

The `ActivatedRouteSnapshot` contains the _future_ route that will be activated and the `RouterStateSnapshot` contains the _future_ `RouterState` of the application, should you pass through the guard check.

If the user is not logged in, you store the attempted URL the user came from using the `RouterStateSnapshot.url` and tell the router to navigate to a login page&mdash;a page you haven't created yet.

> This secondary navigation automatically cancels the current navigation when `checkLogin()` returns `false`.

Reviewer 3: ```suggestion
If the user is not logged in, you store the attempted URL the user came from using the `RouterStateSnapshot.url` and tell the router to redirect to a login page&mdash;a page you haven't created yet.
The router cancels the current navigation and automatically schedules a new one to navigate to the returned `UrlTree`.
```
Something like that. I can't think of a good way to write it...

{@a can-activate-child-guard}

### `CanActivateChild`: guarding child routes

You can also protect child routes with the `CanActivateChild` guard.
The `CanActivateChild` guard is similar to the `CanActivate` guard.

> The key difference is that it runs before  any child route is activated.

Reviewer 3: ```suggestion
The key difference is that it runs before any child route is activated.
```

Generate a `CrisisDetailResolver` service file within the `Crisis Center` feature area.

<code-example language="none" class="code-shell">
  ng generate service crisis-center/crisis-detail-resolver
</code-example>

<code-example path="router/src/app/crisis-center/crisis-detail-resolver.service.1.ts" header="src/app/crisis-center/crisis-detail-resolver.service.ts (generated)"></code-example>

Move the relevant parts of the crisis retrieval logic in `CrisisDetailComponent.ngOnInit()` into the `CrisisDetailResolverService`.
Import the `Crisis` model, `CrisisService`, and the `Router` so you can navigate elsewhere if you can't fetch the crisis.

Be explicit and implement the `Resolve` interface with a type of `Crisis`.

Inject the `CrisisService` and `Router` and implement the `resolve()` method.
That method could return a `Promise`, an `Observable`, or a synchronous return value.

The `CrisisService.getCrisis()` method returns an observable in order to prevent the route from loading until the data is fetched.
The `Router` guards require an observable to `complete`, which means it has emitted all
of its values.

> You use the `take` operator with an argument of `1` to ensure that the Observable completes after retrieving the first value from the Observable returned by the `getCrisis()` method.

Reviewer 3: ```suggestion
You use the `take` operator with an argument of `1` to ensure that the `Observable` completes after retrieving the first value from the Observable returned by the `getCrisis()` method.
```

    </td>
  </tr>

  <tr>
    <td>
      <code>NavigationEnd</code>
    </td>
    <td>

      An [event](api/router/NavigationEnd) triggered when navigation ends successfully.

    </td>
  </tr>

  <tr>
    <td>
      <code>NavigationCancel</code>
    </td>
    <td>

      An [event](api/router/NavigationCancel) triggered when navigation is canceled.
      This can happen when a [Route Guard](#guards) returns false during navigation,
      or redirects by returning a `UrlTree`.

    </td>
  </tr>

  <tr>
    <td>
      <code>NavigationError</code>
    </td>
    <td>

      An [event](api/router/NavigationError) triggered when navigation fails due to an unexpected error.

    </td>
  </tr>

  <tr>
    <td>
      <code>Scroll</code>
    </td>
    <td>

      An [event](api/router/Scroll) that represents a scrolling event.

    </td>
  </tr>
</table>

When you enable the `enableTracing` option, Angular logs these events to the console.
For an example of filtering router navigation events, see the [router section](guide/observables-in-angular#router) of the [Observables in Angular](guide/observables-in-angular) guide.

> ### Router terminology

Reviewer 2: This could be extracted into a separate docs (a router glossary).
<sub>Not necessary on this PR.</sub>

Author: Agreed and it will be moved into the glossary in a follow up PR. Thanks!!

Reviewer 4: @kapunahelewong : I don't think these are glossary-style terms. They are in code font and should probably link to APIs, such as [RouterModule](https://angular.io/api/router/RouterModule). I could see possibly updating the existing [Glossary entry](https://angular.io/guide/glossary#router) to include more terminology.

* There are no parameters for this parent route.
* There is no default for the child route so you need to pick one.
* You're navigating to the `CrisisListComponent`, whose route path is `/`, but you don't need to explicitly add the slash.

Consider the following router link that navigates from the root of the application down to the *Dragon Crisis*:

<code-example path="router/src/app/app.component.3.ts" header="src/app/app.component.ts (Dragon-anchor)" region="Dragon-anchor"></code-example>

* The first item in the array identifies the parent route (`/crisis-center`).
* There are no parameters for this parent route.
* The second item identifies the child route details about a particular crisis (`/:id`).
* The details child route requires an `id` route parameter.
* You added the `id` of the *Dragon Crisis* as the second item in the array (`1`).
* The resulting path is `/crisis-center/1`.

You could also redefine the `AppComponent` template with *Crisis Center* routes exclusively:

<code-example path="router/src/app/app.component.3.ts" header="src/app/app.component.ts (template)" region="template"></code-example>

> In sum, you can write applications with one, two or more levels of routing.

Reviewer 3: ```suggestion
In summary, you can write applications with one, two, or more levels of routing.
```
